File: backend/src/models/AuditLog.ts
import mongoose, { Schema } from 'mongoose';
import { AuditLog as IAuditLog } from '@/types';

const auditLogSchema = new Schema({
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: [true, 'User reference is required']
  },
  patient: {
    type: Schema.Types.ObjectId,
    ref: 'Patient',
    required: [true, 'Patient reference is required']
  },
  accessType: {
    type: String,
    required: [true, 'Access type is required'],
    enum: ['regular', 'emergency', 'consent'],
    default: 'regular'
  },
  action: {
    type: String,
    required: [true, 'Action is required'],
    enum: ['view', 'create', 'update', 'delete']
  },
  details: {
    type: String,
    required: [true, 'Details are required'],
    trim: true,
    maxlength: [1000, 'Details cannot exceed 1000 characters']
  },
  accessTime: {
    type: Date,
    required: [true, 'Access time is required'],
    default: Date.now
  },
  ipAddress: {
    type: String,
    trim: true
  },
  userAgent: {
    type: String,
    trim: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexes
auditLogSchema.index({ user: 1 });
auditLogSchema.index({ patient: 1 });
auditLogSchema.index({ accessType: 1 });
auditLogSchema.index({ accessTime: -1 });
auditLogSchema.index({ createdAt: -1 });

// Virtual for user details
auditLogSchema.virtual('userDetails', {
  ref: 'User',
  localField: 'user',
  foreignField: '_id',
  justOne: true
});

// Virtual for patient details
auditLogSchema.virtual('patientDetails', {
  ref: 'Patient',
  localField: 'patient',
  foreignField: '_id',
  justOne: true
});

export default mongoose.model<IAuditLog>('AuditLog', auditLogSchema);

File: openmrs-modules/patient-passport-core/omod/src/main/java/org/openmrs/module/patientpassportcore/model/AuditLog.java
package org.openmrs.module.patientpassportcore.model;

import org.openmrs.BaseOpenmrsData;
import org.openmrs.Patient;
import org.openmrs.User;

import java.util.Date;

/**
 * Model class for audit logging
 */
public class AuditLog extends BaseOpenmrsData {
	
	private Integer id;
	private User user;
	private Patient patient;
	private String accessType; // regular, emergency, consent
	private String action; // view, create, update, delete
	private String details;
	private Date accessTime;
	private String ipAddress;
	private String userAgent;
	
	public AuditLog() {}
	
	public AuditLog(User user, Patient patient, String accessType, String action, String details) {
		this.user = user;
		this.patient = patient;
		this.accessType = accessType;
		this.action = action;
		this.details = details;
		this.accessTime = new Date();
	}
	
	@Override
	public Integer getId() {
		return id;
	}
	
	@Override
	public void setId(Integer id) {
		this.id = id;
	}
	
	public User getUser() {
		return user;
	}
	
	public void setUser(User user) {
		this.user = user;
	}
	
	public Patient getPatient() {
		return patient;
	}
	
	public void setPatient(Patient patient) {
		this.patient = patient;
	}
	
	public String getAccessType() {
		return accessType;
	}
	
	public void setAccessType(String accessType) {
		this.accessType = accessType;
	}
	
	public String getAction() {
		return action;
	}
	
	public void setAction(String action) {
		this.action = action;
	}
	
	public String getDetails() {
		return details;
	}
	
	public void setDetails(String details) {
		this.details = details;
	}
	
	public Date getAccessTime() {
		return accessTime;
	}
	
	public void setAccessTime(Date accessTime) {
		this.accessTime = accessTime;
	}
	
	public String getIpAddress() {
		return ipAddress;
	}
	
	public void setIpAddress(String ipAddress) {
		this.ipAddress = ipAddress;
	}
	
	public String getUserAgent() {
		return userAgent;
	}
	
	public void setUserAgent(String userAgent) {
		this.userAgent = userAgent;
	}
}

